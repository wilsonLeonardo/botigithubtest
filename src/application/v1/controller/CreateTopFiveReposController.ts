import { RequestHandler } from 'express';
import { CREATED } from 'http-status';
import { ILogger } from '@infrastructure/logger/ILogger';
import { buildErrorInfo } from '@infrastructure/parser/ErrorInfo';
import IGitHubService from '@service/IGitHubService';

export default class CreateTopFiveReposController {
  constructor(private gitHubService: IGitHubService, private logger: ILogger) {}

  public execute: RequestHandler = async (request, response) => {
    try {
      const { programmingLanguages } = request.body;

      const repositories = await this.gitHubService.findAndSaveTopFiveRepos(programmingLanguages);

      return response.status(CREATED).json({ repositories });
    } catch (err: any) {
      const errorInfo = buildErrorInfo(err);

      this.logger.error(`fail to find and save the top five repositories`);

      return response.status(errorInfo.code).json({ error: errorInfo.errorMessage });
    }
  };
}
