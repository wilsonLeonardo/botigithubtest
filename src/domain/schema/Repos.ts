import { Schema } from 'mongoose';

export interface IRepo {
  _id?: number;
  id?: number;
  full_name: string;
  stargazers_count: number;
  language: string;
}

export const Repo: Schema = new Schema<IRepo>(
  {
    _id: {
      type: Number,
    },
    full_name: {
      type: String,
    },
    language: {
      type: String,
    },
    stargazers_count: {
      type: Number,
    },
  },
  {
    timestamps: true,
    toJSON: {
      versionKey: false,
      transform: (_doc, ret) => {
        ret.id = ret._id;
        delete ret._id;
      },
    },
  },
);
