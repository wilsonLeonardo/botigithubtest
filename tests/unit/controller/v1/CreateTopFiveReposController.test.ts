import { Request } from 'express';
import { CREATED, INTERNAL_SERVER_ERROR } from 'http-status';
import { IRepo } from '../../../../src/domain/schema/Repos';
import CreateTopFiveReposController from '../../../../src/application/v1/controller/CreateTopFiveReposController';
import GitHubServiceMock from '../../../mocks/GitHubServiceMock';
import LoggerMock from '../../../mocks/LoggerMock';
import InternalServerError from '../../../../src/domain/exceptions/InternalServerError';
import { repoMock, programmingLanguages } from '../../../mocks/RepoSchemaMock';

describe('CreateTopFiveReposController', () => {
  const githubService = new GitHubServiceMock();
  const logger = new LoggerMock();
  it('should match the snapshot #unit', () => {
    const instance = new CreateTopFiveReposController(githubService, logger);
    expect(instance).toMatchSnapshot();
  });
  it('should create repositories #unit', async () => {
    githubService.findAndSaveTopFiveRepos = jest.fn().mockImplementation((): Promise<Array<IRepo>> => {
      return Promise.resolve(repoMock);
    });
    const instance = new CreateTopFiveReposController(githubService, logger);

    const req = {
      body: {
        programmingLanguages,
      },
    } as unknown as Request;
    const mockJsonResponse = jest.fn();
    const res: any = {
      status: jest.fn().mockImplementation(() => ({
        json: mockJsonResponse,
      })),
    };

    await instance.execute(req, res, jest.fn());
    expect(res.status).toHaveBeenCalledWith(CREATED);
    expect(mockJsonResponse).toHaveBeenCalledWith({ repositories: repoMock });
  });
  it('should return error code when service throws #unit', async () => {
    githubService.findAndSaveTopFiveRepos = jest.fn().mockImplementation((): Promise<Array<IRepo>> => {
      return Promise.reject(new InternalServerError('error message'));
    });
    const instance = new CreateTopFiveReposController(githubService, logger);

    const req = {
      body: {
        programmingLanguages: ['a'],
      },
    } as unknown as Request;
    const mockJsonResponse = jest.fn();
    const res: any = {
      status: jest.fn().mockImplementation(() => ({
        json: mockJsonResponse,
      })),
    };

    await instance.execute(req, res, jest.fn());
    expect(res.status).toHaveBeenCalledWith(INTERNAL_SERVER_ERROR);
    expect(mockJsonResponse).toHaveBeenCalledWith({ error: 'error message' });
  });
});
