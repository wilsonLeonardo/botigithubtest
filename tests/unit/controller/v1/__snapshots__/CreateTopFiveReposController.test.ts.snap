// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTopFiveReposController should match the snapshot #unit 1`] = `
CreateTopFiveReposController {
  "execute": [Function],
  "gitHubService": GitHubServiceMock {},
  "logger": LoggerMock {},
}
`;
